frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -485,42 +474,37 @@ export default ({
           contractID,
           data,
           hooks: {
-            preSendCheck: (message, state) => {
+            preSendCheck: async (message, state) => {
               // NOTE: this preSendCheck does nothing except appending a pending message
               //       temporarily until the uploading attachments is finished
               //       it always returns false, so it doesn't affect the contract state
-              const [, opV] = message.op()
-              const { meta } = opV.valueOf().valueOf()
-
-              temporaryMessage = createMessage({
-                meta,
-                data: { ...data, attachments },
-                hash: message.hash(),
-                height: message.height(),
-                state: this.messageState.contract,
-                pending: true,
-                innerSigningContractID: this.ourIdentityContractId
-              })
-              this.messages.push(temporaryMessage)
```

Comment:
We used to do this (`this.messages.push(temporaryMessage)`)... now we no longer do.

Seems like a big change.

What's going on here? Are we getting rid of the grey unsent-message message? (Why?)

Reply:
This doesn't get rid of the unsent message, or at least isn't supposed to.

See the lines below these deletions, specifically:

```
Vue.set(this.messageState, 'contract', await sbp('chelonia/in/processMessage', message, this.messageState.contract))
```

Instead of manipulating the contract state directly, this uses Chelonia to push the message, which will execute the correct code as defined in the contract.

The previous way, besides relying too much on Chelonia / contract internals, didn't work well with attachments, which now have to cross the browser / SW boundary. This set of changes is meant to address that and simplify the code.

Reply:
> This doesn't get rid of the unsent message, or at least isn't supposed to.

It did appear to have that effect in my tests. Try enabling throttling or set the browser to offline mode and send a message. Compare behavior between `master` and this branch. `master` will show the message but it will be in grey.

Reply:
I've tested this previously and I remember it working. However, I'll test again and see. It needs to be fixed if that's the case.

Reply:
I see, I went offline and messages don't appear at all.

Reply:
I did some additional debugging here and there are two issues. One is easy to solve (which I've solved but haven't committed yet) and the other one is harder.

The easy issue is that outgoing messages have a height of `0`. This is because the height is unimportant and will be set to the correct value when the message is actually sent. Unfortunately, this means that processing messages via Chelonia fails because the signature may be invalid. The fix is setting the height to the maximum possible value instead of 0.

The harder issue is that we rely on the `beforePublish` and `preSendCheck` hooks, which happen after calling `addMessage`. `addMessage` doesn't seem to result in sending a message when offline, and I need to fix this (one reason could be calling `retain`).